src/newtype.rs

```diff
@@ -0,0 +1,57 @@
+use crate::constants::*;
+use rusqlite as sql;
+use rusqlite::types::*;
+use std::fmt;
+
+macro_rules! defnewtype {
+    ($name:ident ($base:ty); $( $field:ident => $value:expr),*) => {
+        #[repr(transparent)]
+        #[derive(Copy, Clone, Eq, PartialEq, Debug)]
+        pub struct $name { value: $base }
+
+        // Display instance reuses instance of underlying type, since it seems
+        // to be impossible to convert identifiers to strings at
+        // macro-expansion time.
+        impl fmt::Display for $name {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(f, "{}", self.value)
+            }
+        }
+
+        impl ToSql for $name {
+            fn to_sql(&self) -> sql::Result<ToSqlOutput> {
+                self.value.to_sql()
+            }
+        }
+
+        impl FromSql for $name {
+            fn column_result(col: ValueRef) -> FromSqlResult<Self> {
+                let inner = FromSql::column_result(col)?;
+                $name::new(inner).ok_or(FromSqlError::InvalidType)
+            }
+        }
+
+        impl $name {
+            pub fn extract(&self) -> $base { self.value }
+            pub fn new(input: $base) -> Option<Self>{
+                $( if input == $value { return Some(Self { value: input }); } )*
+                return None;
+            }
+            $(
+                pub const $field : Self = Self { value: $value };
+            )*
+        }
+    }
+}
+
+defnewtype! {
+    Viewtype(libc::c_int);
+    UNKNOWN => 0,
+    TEXT => DC_MSG_TEXT,
```

Comment:
- use camelcase for enum variants
- remove the constants in favor of this, having this duplicated is unfortunate
- no need for the macro once you use FromPrimitive